Adds EXPERIMENTAL_BUCKET_STORE flag to manage bucket files in an alternative sort order

SirTyson opened the pull request:
# Description

Resolves #3359

This PR adds a new experimental flag `EXPERIMENTAL_BUCKET_STORE` that when set enables building and managing (as-in: merges, garbage collects them, etc) buckets with an alternate sort order. This alternative sort order groups account subentries together with their parent account entry so that they are spatially close together in the `Bucket` file. In the future, we plan to experiment with using the `BucketList` as the primary key-value store instead of an SQL database. This new sort order is required for this experiment to optimize some types of lookups, like when loading all trustlines associated with an account.

When `EXPERIMENTAL_BUCKET_STORE` is set to false, nothing changes. When `EXPERIMENTAL_BUCKET_STORE` is set to true, each `Bucket` will have two files associated with it, one sorted in the traditional order and one sorted in the new experimental sort order. These new files with the experimental sort order are still managed, garbage collected, and merged by the `BucketManager`. However, they are not yet used for anything else. In future updates, the experimental `Bucket` files may be used for key-value lookups with the `BucketList`. For now, since `BucketList` hashes are based off of the older sort order, every `Bucket` object must maintain 2 files each sorted differently. However, future updates could change history publication to use files sorted using the experimental sort order and `Bucket` objects would only need to maintain a single file again.

When `EXPERIMENTAL_BUCKET_STORE` is set, `Bucket` files are resorted in memory with `BucketManager::assumeState()` when downloading buckets from history. While some of the `Bucket` files can be large (currently over 500 mb for lower levels), this in memory sorting occurs only when catching up from history, so it should not be a large performance concern. After this initial in memory resorting, the new experimental files are managed and merged on the background thread along with the traditional, non-experimental `Bucket` files.

This is an on going experiment and it is strongly recommended that `EXPERIMENTAL_BUCKET_STORE` remains set to false (the default setting).

# Checklist
- [X] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [X] Rebased on top of master (no merge commits)
- [X] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [X] Compiles
- [X] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

graydon reviewed (changes requested):
An exciting first cut! I have unfortunately a fairly big laundry list of changes to request. Happy to talk them over if you'd like to have a call.

A couple procedural bits:

- Probably best to get (and be given) earlier design feedback to avoid having to backup/redo as much. Some of this is also on me also for not doing a careful / earlier review, apologies.
- Generally better from a review perspective to split patches like this up into a few separate commits focusing on logically distinct changes. Sometimes impossible -- things depend on one another -- but if you can get into the habit, it makes design review easier also, don't have to piece-together the plan of the thing as much.

A couple broad but shallow bits:

- Generally I think we've already got too many `bool`s flying around this file, I'd like to predicate this work on an `enum class`.
- Generally want to avoid depending-on or inspecting filenames, it's fragile and nonobvious to future readers.
- Generally want to start migrating code / writing new code against `std::filesystem::path` since we have it now.
- Generally want to avoid code duplication when it's not too much hassle. There are limits and to some extent it's a matter of taste where the limit is, but there's a bit more in here than I'm comfortable with.

Then the one big, deep bit:

- I'm not sure about the mapping between Buckets, filenames, and sort-orders of them.
  - I'd like buckets to still have a hash that is the hash of their content, even new ones.
  - I'd like the sort order to just be implied by the protocol version embedded in the `MetaEntry`, not the filename.
  - I think it might work better to have 1 `Bucket` object managing one of 3 possible combinations: just 1 old-protocol file, 1 old + 1 new, or just 1 new.
  - Then `BucketManager` still maps both old and new hashes to the `shared_ptr<Bucket>` that's responsible for both files.
  - The `Bucket` file can have 2 `std::optional<std::filesystem::path>` objects, and "both empty" can subsume the case we currently represent with an empty-string filename / all-zero hash.
  - The `BucketManagerImpl::forgetUnreferencedBuckets` routine will need to adjust its notion of a "singly-referenced `Bucket`" a little bit to accommodate, but it shouldn't be too bad.

Some of this is me looking/thinking ahead about how we're going to do the longer-term transition of buckets on disk (and in archives) from old-protocol to new-protocol when we finally cut over. It'd be nice if we could just switch to a `BucketList` full of equivalent-but-differently-named files that happen to be new-protocol instead of the old, and drop/dereference the old ones.

It's totally possible that this plan contradicts part of the plan we discussed earlier, and/or is impossible for structural reasons I can't quite see at the moment. Would be happy to discuss!

SirTyson commented:
I think that all sounds reasonable, except I'd like to discuss this point a little more with @MonsieurNicolas:

> I'd like the sort order to just be implied by the protocol version embedded in the `MetaEntry`, not the filename.

I originally extended the XDR BucketMetadata object to include a new flags field where you could mark the sort order, but Nico mentioned that changing this struct would be a pretty big lift given it would be changing the format of published history. I don't know if using the protocol version is the right solution either. A single Bucket object could have 2 valid files backing it in the old protocol + new protocol configuration, but each file would be a different protocol version. At least to me, it feels weird that two different protocol versions could be valid at the same time, although maybe this is more common than I'm thinking. I'm also concerned in the case that we update the new sort function later, then we're in a weird place where protocol version 18 (the original sort order) and 20 (the new sort order V2) are valid, but version 19 is out of date. I may be overthinking the issue though.

MonsieurNicolas commented:
> I think that all sounds reasonable, except I'd like to discuss this point a little more with @MonsieurNicolas:
> 
> > I'd like the sort order to just be implied by the protocol version embedded in the `MetaEntry`, not the filename.
> 
> I originally extended the XDR BucketMetadata object to include a new flags field where you could mark the sort order, but Nico mentioned that changing this struct would be a pretty big lift given it would be changing the format of published history. I don't know if using the protocol version is the right solution either. A single Bucket object could have 2 valid files backing it in the old protocol + new protocol configuration, but each file would be a different protocol version. At least to me, it feels weird that two different protocol versions could be valid at the same time, although maybe this is more common than I'm thinking. I'm also concerned in the case that we update the new sort function later, then we're in a weird place where protocol version 18 (the original sort order) and 20 (the new sort order V2) are valid, but version 19 is out of date. I may be overthinking the issue though.

yeah in this context, reason I wanted to avoid changing the xdr was that it opens up new attack vectors/ways to expose the wrong bucket type and at this stage this should be "invisible" from a protocol point of view.

in this iteration it may overcomplicate matters to try to generate buckets on disk that are "the new format": when we finally make this a protocol change, we would have to convert them to use the new protocol version anyways (for the reason @SirTyson described above) so it would not just be a matter of "switching hash in the bucket list".
Also, I think that if we were to use a hash for those alternate buckets, it creates a new problem that is how to "re-attach" those buckets when we restart as we can't just guess the filename of the other file when dealing with "old sort order" buckets.

SirTyson commented on src/bucket/BucketInputIterator.cpp:
```diff
@@ -87,7 +87,8 @@ BucketInputIterator::getMetadata() const
 
-BucketInputIterator::BucketInputIterator(std::shared_ptr<Bucket const> bucket)
+BucketInputIterator::BucketInputIterator(std::shared_ptr<Bucket const> bucket,
+                                         std::string const& inFilename)
```
I've changed the internals of `Bucket` to have an optional filename/hash for each sort order. This means that a bucket can now have just the old sort order, just the new sort order, both, or neither (representing the default, empty `Bucket` object). However, I still have `Bucket::getFilename(BucketSortOrder) -> fs::path` and `Bucket::getHash(BucketSortOrder) -> Hash`. Currently the codebase treats empty string and `Hash{}` as null values, so I have the functions return those values if the underlying class member is `std::nullopt`. Is it worth refactoring the codebase to support actual optionals instead of these null values? It's a fairly large refactor so I wanted to double check before doing it. Other than having to copy values instead of returning a reference, I don't see any other drawback for optional values. However I wanted to make sure that there wasn't some advantage or edge case I was missing making an empty `Hash` better than an `std::optional<Hash>`.

graydon replied:
I think it's a worthwhile refactor to try in the future -- along with ever-more migrations from std::strings to std::filesystem::path -- but I wouldn't gate this change on it, you've got enough going on in here already, and IIRC we special-case the empty / zero hash in a number of places in the codebase, not just buckets.

graydon replied:
Actually as I read it and fiddle around with the code I'm increasingly nervous that we're conflating "empty hash" with important state transitions here. It might be worth refactoring to make the optionals explicit in more places. I'm going to give it a try to see how terrible it is, might add that last bit if it seems ok.

SirTyson commented on src/bucket/LedgerCmp.h:
```diff
@@ -88,62 +91,198 @@ struct LedgerEntryIdCmp
-                    "Malformed bucket: unexpected non-INIT/LIVE/DEAD entry.");
-            }
-            if (bty == LIVEENTRY || bty == INITENTRY)
+            if (bty != DEADENTRY)
```
I don't believe it is. I've refactored and moved the code around a bit now that we have two different compare functions sharing the common lifecycle-state comparison code, but the actual logic should be the same.

SirTyson commented on src/bucket/LedgerCmp.h:
```diff
@@ -88,62 +91,198 @@ struct LedgerEntryIdCmp
+ */
+struct BucketEntryIdCmp
+{
+    virtual bool
```
I'm still not sure the best way to go about the compare functions. I need the two compare functions to have the same type to be able to pass them as parameters to functions. To avoid class inheritance and virtual dispatch, I ended up with a `typedef` of the compare function prototype with `std::function`, and changed the two compare functors to be a single, templated function. The template allows you to specify an additional type parameter while still maintaining the `std::less` prototype, and the `typedef` lets you pass the functions as a parameter. It works and seems syntactically sound, but I still feel like it's not the best interface.

graydon replied:
Yeah, reading it I'm not sure it's a net improvement or loss, or if it'd help to make things even-more-templated / statically type-based; but there's no point going around in circles on it, I'm fine with this version for now.

graydon reviewed (changes requested):
Looking good! Couple more items to refine (as discussed in meeting today):

- Split the `getHash` function into `getHash(type)` and `getPrimaryHash()` where the former has no default argument and the latter returns its "highest-protocol" hash, which is the by-account hash if it exists, else the by-type hash. Transition callers to `getPrimaryHash` in cases where the actual hash used will vary dynamically (eg. mergekey calculation, probably bucket manager adoption?) and `getHash(type)` when there is a definite type required by the calling context (such as when calculating a hash for the bucketlist when closing a ledger at a particular protocol version). Possibly do something similar to `getFilename` if there are both kinds of callers for that too.
- Maybe: return an `optional<Hash>` and `optional<path>` in the type-based accessors -- this is less-obviously important (or even correct) to me at this point but I've got a little ways into trying it in a change I'll push here, and I'm slightly leaning towards trying to include it since we have optionals in the data model already. But it might also be the case that continuing to have the empty hash as a valid sentinel is less disruptive / that using optionals doesn't buy much clarity, in which case just throw that bit out -- I think clarifying the first issue above is the more important one.

SirTyson commented:
> Looking good! Couple more items to refine (as discussed in meeting today):
> 
> * Split the `getHash` function into `getHash(type)` and `getPrimaryHash()` where the former has no default argument and the latter returns its "highest-protocol" hash, which is the by-account hash if it exists, else the by-type hash. Transition callers to `getPrimaryHash` in cases where the actual hash used will vary dynamically (eg. mergekey calculation, probably bucket manager adoption?) and `getHash(type)` when there is a definite type required by the calling context (such as when calculating a hash for the bucketlist when closing a ledger at a particular protocol version). Possibly do something similar to `getFilename` if there are both kinds of callers for that too.
> * Maybe: return an `optional<Hash>` and `optional<path>` in the type-based accessors -- this is less-obviously important (or even correct) to me at this point but I've got a little ways into trying it in a change I'll push here, and I'm slightly leaning towards trying to include it since we have optionals in the data model already. But it might also be the case that continuing to have the empty hash as a valid sentinel is less disruptive / that using optionals doesn't buy much clarity, in which case just throw that bit out -- I think clarifying the first issue above is the more important one.

I've gone ahead and removed the default type parameter for `getHash()` and added the `getPrimaryHash()` function. I think it makes sense for `getHash()` to return a `std::optional`, especially since it's not called very often now that `getPrimaryHash()` is available. However, for `getPrimaryHash()`, I think it's best to return either the valid hash, or the empty sentinel hash. Since `getPrimaryHash()` is used as an internal identifier, the functions that call it always expect to receive a hash back. There's a lot of logic where primary hashes are compared, and it's easier to compare a (potentially) empty sentinel hash than to plumb `std::optional` through everything. Also, adding empty hashes to the `BucketList` hasher actually affects the outcome of the hash. This means that we'll have to rely on using an empty sentinel hash to have consistent hashes no matter what, so the `std::optional` divide between `getHash()` and `getPrimaryHash()` makes sense to me.

There were also a few cases in the code where functions would call `getHash()` and check to see if the return value was the empty hash to see if the `bucket` object was empty or not. This is used in places like `FutureBucket`, where the associated `Bucket` object may be empty if the underlying file is not finished being sorted. I've added the function `Bucket::isEmpty()` for these cases.

Finally, it's important to note that the primary hash returned by `getPrimaryHash` may not be the same hash used for ledger hashes. Because of this, I've added a constant `LEDGER_HASH_SORT_ORDER` to specify the correct history type.

graydon commented on src/main/test/ApplicationUtilsTests.cpp:
```diff
@@ -106,11 +106,13 @@ TEST_CASE("offline self-check works", "[applicationutils][selfcheck]")
+                              .getBucketList()
+                              .getLevel(0)
+                              .getCurr()
+                              ->getFilename(BucketSortOrder::SortByType);
```
This will stop working when we transition to the new type. Maybe we need a `getPrimaryFilename` as well, for such cases?

graydon commented on src/main/ApplicationUtils.cpp:
```diff
@@ -421,7 +421,9 @@ mergeBucketList(Config cfg, std::string const& outputDir)
 
     using std::filesystem::path;
-    path bpath(bucket->getFilename());
+    auto pathOp = bucket->getFilename(BucketSortOrder::SortByType);
```
This will also stop working when we switch.

graydon commented on src/history/StateSnapshot.cpp:
```diff
@@ -149,7 +149,8 @@ StateSnapshot::differingHASFiles(HistoryArchiveState const& other)
         auto b = mApp.getBucketManager().getBucketByHash(hexToBin256(hash));
         releaseAssert(b);
-        addIfExists(std::make_shared<FileTransferInfo>(*b));
+        addIfExists(std::make_shared<FileTransferInfo>(
```
FileTransferInfo I think will need to use `getPrimaryFile` or else it'll break on switch, no?

graydon reviewed (changes requested):
Close! I mentioned a couple things in the line-item comments, roughly:

  - I think the hash-that-is-hashed-into-a-ledger is not static, but will vary depending on the ledger that is closing / acquiring the hash, so has to be a dynamic argument.
  - I think there are a bunch of contexts here that want to use something like a "primary filename" that works the same way the primary hash does -- picks the newest one attached to the bucket object. But .. now that I say that aloud .. I think that might also not be quite right either; for example maybe the FileTransferInfo transferring to archives .. might actually need to know which ledger version it's being snapshotted _from_, in order to know whether to capture the old-order or new-order bucket file.

SirTyson commented:
The newest commit adds a `HashID` type to work as a consistent internal `Bucket` identifier. Now that `Bucket` is backed by two files with different hashes, it was difficult to make sure the same hash was used universally to identify that `Bucket`. `HashID` is a weak type wrapper around `Hash`, but makes sure that a single, consistent `Hash` is used for the same bucket. Now whenever a `Bucket` needs to be referenced, we use the `HashID`. When an individual file is referenced, like in the HistoricalArchiveState, regular `Hash` are still used to identify them. Ideally there would be a strong divide between `HashID` and `Hash`, but unfortunately that is not possible, and sometimes it is still necessary to turn a `HashID` into a `Hash` or vice versa. Several objects, such as `HistoryArchiveState` and `FutureBucket`, do not manage `Bucket` objects, but actually manage individual files that later become `Bucket`s. This means that these objects often convert `Hash` to `HashID`. These conversions are tricky, and these type of bugs took me a while to track down. I think it is correct, but I still want to add a few more unit tests. The final result uses many more `Hash` than I would like, but the `HashID` type is still very helpful for keeping track of buckets in `MergeKey` and `BucketManager`.

marta-lokhova commented:
This is cool, thank you for putting this together! I started reviewing the change in detail, but kept coming back to one design aspect, which I think impacts the rest of my comments, so I'll outline it first. 
If I understand this correctly:
* before this change, nodes manage a single version of buckets, sorted by types
* when this change is merged and released, nodes will begin managing two versions of a bucket, sorted by type and by accounts. However, the other version of buckets won't be used anywhere.
* Starting protocol 20, nodes will switch to a different bucket sort order, at which point it will impact ledger hashes (as nodes will begin producing new buckets exclusively, and switch older buckets to the new version)

Is that right? A had a few questions, if so:

* We're changing how the ledger hash is computed, so I think we need to write a CAP for these changes? Are we ok merging stuff into master that enforces certain behavior without having an accepted CAP?
* More importantly, do we need to maintain two bucket versions before the protocol upgrade? I wonder if nodes could start generating new buckets at the protocol switch. Kind of like what we did in CAP-0020 and CAP-0025. This should be ok perf-wise, because nodes are only blocked on level 0, and they can gradually merge/resort buckets in the background. I understand that we can't actually take advantage of k-v store without the bucketlist being resorted completely, so I wonder if this is a matter of protocol upgrade timing? Hypothetically, if we had a protocol upgrade at the same time as we planned to release this change (with old protocol), we'd achieve the same result? The reason I'm asking is because it seems like we could simplify the code quite a bit if we didn't have to manage two files, as we could just reuse a lot of the bucket code that we already have in master, and we wouldn't need to worry about the archives/catchup flow. I think this would eliminate having two different valid versions of the same bucket as well.

Sorry for asking all these questions so late in the process, but wanted to see what you guys think.

MonsieurNicolas commented:
Replying to @marta-lokhova 's question.
For the "does this need a CAP?", eventually yes, what we're doing now is experiment with the changes before writing a formal CAP (which is why we can enable it via an "experimental" flag in the config).
That said, we should probably have a short blurb that explains what happens during the protocol upgrade (and choices that were made) somewhere (this was not in the original scope of this PR, but we saw that it made more sense to prepare for it instead of writing code that ends up being invalidated right after): either added to the description of this PR or pushed as a "very rough draft" CAP (that we can edit and refine over time). I'll let @SirTyson decide on what he prefers.

SirTyson commented:
Summarizing the last few days of conversation and replying to Graydon's email thread here for visibility.

It looks like we have 2 possible solutions, either we let the new sort order gradually make its way down the bucket list, or we switch the entire bucket list to the new sort order at the same time. The "trickle down" method has a straightforward implementation, but there are 3 possible ways to switch the entire bucket list over to the new sort order:
1. The current two files per bucket approach
2. Maintain two bucket lists, one for each sort order, with each bucket only having a single file
3. Each bucket has a single file and there is  a single bucket list. Before the p rotocol upgrade, the background merge thread resorts bucket files to the new sort order, but neither the bucketlist or the bucket objects know about the newly sorted files until the protocol is updated.

Initially, we agreed that the trickle down approach was best. It required the least code change, and the entire bucket list would be updated to the new sort order in about 60 days. 

However, after talking with Nico, I think we may have missed another issue with the trickle down approach. If we go with trickle down buckets, we are going to have to keep the  SQL code for history replay. Since the older buckets will all be sortByType, for certain types of queries we would have to do inefficient linear lookups. For example, to get all the trustlines associated with a given account, we would have to traverse the entire bucket list with a linear search if the bucket list is not in the new sort order. We would have to maintain the SQL code to make these queries even remotely efficient. Alternatively, we could resort buckets during history catchup so we could leverage the more efficient lookups, but then we're in the same place as before as far as managing two different bucket sort orders.

I would like to gt rid of as much SQL as possible, so I'm thinking that resorting the entire bucket list at once is still best. Essentially, we need a way to maintain two bucket sort orders at once, not only for the initial update period, but also for history replay. Looking at the three options above, I don't think the approach with single file buckets and a single bucket list is possible (option 3 above). While this would make the transition easier, there is no way to maintain two usable sort orders at once, as the bucket list either references the old sort order files or the new ones but never both.

For the remaining two options, I think the two bucket list approach would be easier. Currently with the two files per bucket approach, whenever a bucket is referenced, we need to know whether to use the sortByType or sortByAccount hash. This gets complicated fast, as many parts of the code like HAS and FutureBucket reference buckets by hash.

The best option looks like the two bucket list approach. We can keep the experimental flag so that validators can resort the bucketlist prior to the actual upgrade. We can also look into the external sort for resorting to keep the memory impact low. For history replay, for buckets with protocol version < 20, we maintain a sortByType bucketlist as well as resorting the bucket to a parallel sortByAccount bucket list. We then can use the sortByType bucketlist for hashes and the sortByaccount bucketlist for lookup, allowing us to get away from SQL. when the protocol version is >= 20, we can then drop the sortByType bucketlist and only keep the sortByAccount bucketlist for both hashes and lookup. This means that we only have to maintain two lists during history catchup and the initial update period but can use just one bucket list for normal operation. 

The current implementation essentially already does all this, just at the bucket level instead of the bucketlist level. While there are some issues that two bucket lists introduce (such as making sure the two lists are consistent w.r.t. merges), I think it will be much cleaner than the two file per bucket approach.

Thoughts @marta-lokhova @graydon? There's been a lot of back and forth so I'd like to get a group consensus before spending too much time on anything.

graydon commented:
Replaying all of history is a fairly niche application that only happens when people are regenerating archives or when we're doing correctness checks in our nightly CI cycles -- both of which are _offline_ operations that are not particularly performance sensitive, so long as the slow event isn't ubiquitous.

IOW I think we should carefully consider whether the set of problematic operations, their frequency and the size of the ledger at the times when they occur actually justify this concern. If they're not super frequent and super slow at the times they occur, I think we should just accept that replay will have a few O(N) stutters in it.

A periodic stall of a few seconds or even minutes here and there through a history replay (which takes days) is not a big deal. It's a big deal to have in the ledger-close loop for new txs arriving, since anything O(N) a tx can trigger _online_ would be a DoS vector; that's why we're pursuing this resorting thing in advance of using the BL online. But it's not a big deal if it's only a thing that causes little stutters to the handful of people who replay history _offline_.

I'm actually struggling to find the full set of operations that are a problem. You mention "loading all TLs associated with an account" but as far as I can tell that doesn't .. seem to be something we currently do (though I may be mistaken, I can't find such an access pattern in the TL SQL code). @MonsieurNicolas has mentioned some operations of concern before but I am not certain which they are. There are no secondary indexes on the TL table so I don't _think_ it actually has to do with that. There are also no secondary indexes on the accounts, accountdata, or claimablebalance tables. There's an index on offers by seller, which is used by the authorization-revocation path of SetTrustLineFlags -- if a TL is deauthorized, we delete all the offers made by the TL trustor. But the orderbook is trivially small through most of history, I don't _think_ auth-revocation is actually very frequent, and I think we're going to have to keep the orderbook (or a significant index of it) in memory in this new scheme anyways. So if that's the only index we're compensating for, and only in offline replay, I think we should just accept the cost of doing an O(N) operation on it here and there.

Interested to hear others' input. Especially if you know of other operations we'd take a hit on.

MonsieurNicolas commented:
> Replaying all of history is a fairly niche application that only happens when people are regenerating archives or when we're doing correctness checks in our nightly CI cycles -- both of which are _offline_ operations that are not particularly performance sensitive, so long as the slow event isn't ubiquitous.

I agree. This depends on how bad it is: CI may not sound that important, but if we can't validate builds in a timely fashion it's a tax we will pay "forever after" (ie: it's the area under the curve that matters). In this case if we can't parallelize enough.

> 
> IOW I think we should carefully consider whether the set of problematic operations, their frequency and the size of the ledger at the times when they occur actually justify this concern. If they're not super frequent and super slow at the times they occur, I think we should just accept that replay will have a few O(N) stutters in it.

Doing a test on a machine using the `diag-bucket-stats` command on a few buckets, it looks like we can scan about 220MB of ledger data per second (tested this in AWS on our target SKU, with a local NVMe drive), so for a ledger of today's size (9.3GB) a full scan takes about 45 seconds.


> 
> A periodic stall of a few seconds or even minutes here and there through a history replay (which takes days) is not a big deal. It's a big deal to have in the ledger-close loop for new txs arriving, since anything O(N) a tx can trigger _online_ would be a DoS vector; that's why we're pursuing this resorting thing in advance of using the BL online. But it's not a big deal if it's only a thing that causes little stutters to the handful of people who replay history _offline_.

It depends - the theoretical stall would be in the order of 12 hours per ledger using today's ledger size. This number of course can be quite different by the time we have a fully sorted bucket list.

> 
> I'm actually struggling to find the full set of operations that are a problem. You mention "loading all TLs associated with an account" but as far as I can tell that doesn't .. seem to be something we currently do, or rather doesn't seem to be something that would even be a problem if we did do. TLs have (account,asset) as their primary key, and we do loads by that pair alone.
> 
> I may be mistaken though, I just can't find a problematic access pattern in the TL SQL code. @MonsieurNicolas has mentioned some operations of concern before but I am not certain which they are. There are no secondary indexes on the TL table so I don't _think_ it actually has to do with that. There are also no secondary indexes on the accounts, accountdata, or claimablebalance tables. There's an index on offers by seller, which is used by the authorization-revocation path of SetTrustLineFlags -- if a TL is deauthorized, we delete all the offers made by the TL trustor. But the orderbook is trivially small through most of history, I don't _think_ auth-revocation is actually very frequent, and I think we're going to have to keep the orderbook (or a significant index of it) in memory in this new scheme anyways. So if that's the only index we're compensating for, and only in offline replay, I think we should just accept the cost of doing an O(N) operation on it here and there.

If we didn't have such queries we would not have to resort buckets :)

We actually ran into perf issues because of this in captive core in the in memory version (the gitst that keeps on giving).

What you are looking for is `InMemoryLedgerTxn::getPoolShareTrustLinesByAccountAndAsset` (for the in memory version), for SQL, we get the lookup by account for free because we have `PRIMARY KEY  (accountid, asset))`.


> 
> Interested to hear others' input. Especially if you know of other operations we'd take a hit on.

graydon commented:
> CI may not sound that important, but if we can't validate builds

I'm not talking about the inability to validate builds; I'm talking about stalls of a few tens of seconds here and there on single shards of our 128-way parallel replay that takes a whole day anyways. Quantities matter here, and if the quantity is "lost in the noise" (as I suspect) then it's fine. So let's get into quantities.

> for a ledger of today's size (9.3GB) a full scan takes about 45 seconds

That's worst case size as of today, and assuming you scan the whole ledger. Through most of history the ledger is much much smaller, so any scan will be much faster. And as far as I can tell (see below) there's no need to scan the whole ledger.

> the theoretical stall would be in the order of 12 hours per ledger using today's ledger size

If you got to "12 hours" by taking 45 seconds and multiplying by 1000 ops in a ledger, then again you're doing worst case analysis where it's not relevant: we want to know what the actual distribution of bad ops is in our actual history.

> If we didn't have such queries we would not have to resort buckets

Ok well .. maybe we don't? Let's back up a step further. We're resorting buckets in order to cluster all LEs associated with a single _account_ in a contiguous physical extent in a bucket. Whereas today (in the existing sort order) they're clustered by type first. But the next-most-significant elements of [the existing `LedgerKey` sort order for, say, `TrustLine`, is the `(account, asset)` pair](https://github.com/stellar/stellar-core/blob/master/src/bucket/LedgerCmp.h#L70-L72). So if we continue with the prototype "binary search index" in memory over buckets in the existing sort order, we can get all the _TLs_ associated with a given (account, asset) pair by doing a range query on `(TRUSTLINE, account, asset)` and reading that range. It just won't be physically close to _other_ non-TL LEs that are associated with that account. The resorting is to get _all_ the LEs associated with an account clustered. But they're already clustered by account inside a single LE type.

But maybe that's ok? Part of the motivation for all this work was the thought that we'd be bulk loading all the LEs associated with an account for smart contracts -- based on assumptions we made about direct interop, and data being subkeyed strictly by account, and so forth -- but the assumptions there might not even hold anymore. And even they do still hold, and we do want to resort, it's not clear to me that this specific op -- `getPoolShareTrustLinesByAccountAndAsset` -- actually can't be done efficiently on the existing sort order.

(Or that if it can't be, that it's frequent enough and expensive enough to warrant concern on offline replay)

> What you are looking for is InMemoryLedgerTxn::getPoolShareTrustLinesByAccountAndAsset

Ok so that is slow because it scans the whole ledger _in memory_ and it has to do that because the in-memory ledger is a hashtable. On disk it's sorted, just not optimally.

But, ok, temporarily setting aside the question above -- do we even _have_ a slow op to worry about -- it seems to me like we only hit this path from `removeOffersAndPoolShareTrustLines` which, again, is only the revoke-auth path of `SetTrustLineFlags`. I just mucked around with hubble and it's unclear to me how often auth _revocation_ actually happens (something like 1 in 20 or so flag-sets from a casual browse of the data -- the actual query needed isn't supported by bigquery) and as far as I can tell there are a total of 40k `SetTrustLineFlags` ops in the _entire_ history.  So ballpark about 2000 total slow ops in all of history, and we replay 128-way parallel. So each worker gets hit with 15 slow ops, each of which is something like 10 seconds over most of history (before the ledger gets huge) and is _at worst_ 45 seconds if all other assumptions here fail and they all happen on giant current ledger.

I just don't see this is worth worrying about.

And I think we should also plausibly revisit "is any of this resorting necessary at all, given current plans". Maybe it's not! The existing order does have a subkey order that supports range queries within an LE type. Just not across all of them.

SirTyson commented:
> > CI may not sound that important, but if we can't validate builds
> 
> I'm not talking about the inability to validate builds; I'm talking about stalls of a few tens of seconds here and there on single shards of our 128-way parallel replay that takes a whole day anyways. Quantities matter here, and if the quantity is "lost in the noise" (as I suspect) then it's fine. So let's get into quantities.
> 
> > for a ledger of today's size (9.3GB) a full scan takes about 45 seconds
> 
> That's worst case size as of today, and assuming you scan the whole ledger. Through most of history the ledger is much much smaller, so any scan will be much faster. And as far as I can tell (see below) there's no need to scan the whole ledger.
> 
> > the theoretical stall would be in the order of 12 hours per ledger using today's ledger size
> 
> If you got to "12 hours" by taking 45 seconds and multiplying by 1000 ops in a ledger, then again you're doing worst case analysis where it's not relevant: we want to know what the actual distribution of bad ops is in our actual history.
> 
> > If we didn't have such queries we would not have to resort buckets
> 
> Ok well .. maybe we don't? Let's back up a step further. We're resorting buckets in order to cluster all LEs associated with a single _account_ in a contiguous physical extent in a bucket. Whereas today (in the existing sort order) they're clustered by type first. But the next-most-significant elements of [the existing `LedgerKey` sort order for, say, `TrustLine`, is the `(account, asset)` pair](https://github.com/stellar/stellar-core/blob/master/src/bucket/LedgerCmp.h#L70-L72). So if we continue with the prototype "binary search index" in memory over buckets in the existing sort order, we can get all the _TLs_ associated with a given (account, asset) pair by doing a range query on `(TRUSTLINE, account, asset)` and reading that range. It just won't be physically close to _other_ non-TL LEs that are associated with that account. The resorting is to get _all_ the LEs associated with an account clustered. But they're already clustered by account inside a single LE type.
> 
> But maybe that's ok? Part of the motivation for all this work was the thought that we'd be bulk loading all the LEs associated with an account for smart contracts -- based on assumptions we made about direct interop, and data being subkeyed strictly by account, and so forth -- but the assumptions there might not even hold anymore. And even they do still hold, and we do want to resort, it's not clear to me that this specific op -- `getPoolShareTrustLinesByAccountAndAsset` -- actually can't be done efficiently on the existing sort order.
> 
> (Or that if it can't be, that it's frequent enough and expensive enough to warrant concern on offline replay)
> 
> > What you are looking for is InMemoryLedgerTxn::getPoolShareTrustLinesByAccountAndAsset
> 
> Ok so that is slow because it scans the whole ledger _in memory_ and it has to do that because the in-memory ledger is a hashtable. On disk it's sorted, just not optimally.
> 
> But, ok, temporarily setting aside the question above -- do we even _have_ a slow op to worry about -- it seems to me like we only hit this path from `removeOffersAndPoolShareTrustLines` which, again, is only the revoke-auth path of `SetTrustLineFlags`. I just mucked around with hubble and it's unclear to me how often auth _revocation_ actually happens (something like 1 in 20 or so flag-sets from a casual browse of the data -- the actual query needed isn't supported by bigquery) and as far as I can tell there are a total of 40k `SetTrustLineFlags` ops in the _entire_ history. So ballpark about 2000 total slow ops in all of history, and we replay 128-way parallel. So each worker gets hit with 15 slow ops, each of which is something like 10 seconds over most of history (before the ledger gets huge) and is _at worst_ 45 seconds if all other assumptions here fail and they all happen on giant current ledger.
> 
> I just don't see this is worth worrying about.
> 
> And I think we should also plausibly revisit "is any of this resorting necessary at all, given current plans". Maybe it's not! The existing order does have a subkey order that supports range queries within an LE type. Just not across all of them.

From a practical perspective, the trickle down buckets are not very complicated to implement. For the history replay concerns, it might be a good idea just to make a rough draft trickle down version and get a measurement or two. Especially since the inefficient look ups would all happen on historical data anyway and no new ledgers after protocol 20, we could quantify the change and know it wouldn’t get worse.

I think it’s a good idea to also revisit the sorting order briefly as Graydon mentioned. This is probably a question best answered by you all, as I don’t have a good sense on how smart contract will change our data access patterns (if at all).

MonsieurNicolas commented:
> I think it’s a good idea to also revisit the sorting order briefly as Graydon mentioned. This is probably a question best answered by you all, as I don’t have a good sense on how smart contract will change our data access patterns (if at all).

I think we already have the answer for smart contracts: we are not going to have range queries, and even if we want to add some in the future, what we can do is make sure to pick `LedgerKey` for "smart data" that are clustered together (and right now they are: we are doing `accountID | contractID | ...`).

As far as how it would go using the existing sort order, it's plausible that it would work OK.
I think the difference is that we would need to perform one range query per bucket entry type per level, so 3 times (live/init/dead) as many lookups than with the alternate sort order.
Would we lose anything else?

MonsieurNicolas commented:
moving away from this approach
